protocols/autonat/src/v2/server/handler/dial_back.rs

```diff
@@ -0,0 +1,136 @@
+    type OutboundProtocol = ReadyUpgrade<StreamProtocol>;
+    type InboundOpenInfo = ();
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(DeniedUpgrade, ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        if let Poll::Ready(result) = self.outbound.poll_unpin(cx) {
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                result
+                    .map_err(|timeout| io::Error::new(io::ErrorKind::TimedOut, timeout))
+                    .and_then(identity),
+            ));
+        }
+        if let Some(cmd) = self.pending_nonce.take() {
+            self.requested_substream_nonce = Some(cmd);
+            return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(ReadyUpgrade::new(DIAL_BACK_PROTOCOL), ()),
+            });
+        }
+        Poll::Pending
+    }
+
+    fn on_behaviour_event(&mut self, _event: Self::FromBehaviour) {}
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::FullyNegotiatedOutbound(FullyNegotiatedOutbound {
+                protocol, ..
+            }) => {
+                if let Some(cmd) = self.requested_substream_nonce.take() {
+                    if self
+                        .outbound
+                        .try_push(perform_dial_back(protocol, cmd))
+                        .is_err()
+                    {
+                        tracing::warn!("Dial back dropped, too many requests in flight");
+                    }
+                } else {
+                    tracing::warn!("received dial back substream without nonce");
+                }
+            }
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: StreamUpgradeError::NegotiationFailed | StreamUpgradeError::Timeout,
+                ..
+            }) => {
+                if let Some(cmd) = self.requested_substream_nonce.take() {
+                    let _ = cmd.back_channel.send(Err(DialBackRes::DialBackErr));
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+async fn perform_dial_back(
+    stream: impl AsyncWrite + Unpin,
+    DialBackCommand {
+        nonce,
+        back_channel,
+        ..
+    }: DialBackCommand,
+) -> io::Result<()> {
+    let res = dial_back(stream, nonce)
+        .await
+        .map_err(|_| DialBackRes::DialBackErr)
+        .map(|_| ());
+    futures_time::task::sleep(futures_time::time::Duration::from_millis(100)).await;
```

Comment:
Let's add a comment here, explaining why we are doing this. Plus, I think we should raise this issue with the spec!

Reply:
explained it and mentioned it in the pr.